src/main/java/ch/njol/skript/expressions/ExprIndicesOfX.java

```diff
@@ -0,0 +1,200 @@
+		);
+	}
+
+	private IndexType type;
+	private boolean position;
+	private Expression<?> value;
+	private Expression<?> objects;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (exprs[1].isSingle()) {
+			Skript.error("'" + exprs[1].toString(null, false)
+					+ "' can only ever have one value at most, thus the 'indices of x in list' expression has no effect.");
+			return false;
+		}
+
+		if (!(exprs[1] instanceof Variable<?>) && matchedPattern == 0) {
+			Skript.error("'" + exprs[1].toString(null, false)
+					+ "' is not a list variable. You can only get the indices of a list variable.");
+			return false;
+		}
+
+		position = matchedPattern == 1;
+		objects = LiteralUtils.defendExpression(exprs[1]);
+		type = IndexType.values()[parseResult.mark];
+		value = LiteralUtils.defendExpression(exprs[0]);
+
+		return LiteralUtils.canInitSafely(objects, value);
+	}
+
+	@Override
+	protected Object @Nullable [] get(Event event) {
+		Object value = this.value.getSingle(event);
+		if (value == null) {
+			if (this.position)
+				return new Integer[0];
+			return new String[0];
+		}
+
+		List<Object> indices = new ArrayList<>();
+
+		int position = 1;
+		if (objects instanceof Variable<?> list) {
+			//noinspection unchecked
+			Map<String, Object> variable = (Map<String, Object>) list.getRaw(event);
+			if (variable == null) {
+				if (this.position)
+					return new Integer[0];
+				return new String[0];
+			}
+
+			for (Map.Entry<String, Object> entry : variable.entrySet()) {
+				Object entryValue = entry.getValue();
+				// the value of {foo::1} when {foo::1::bar} is set is a map with a null key of the value {foo::1}
+				if (entryValue instanceof Map<?, ?> map)
+					entryValue = map.get(null);
+
+				if (entryValue.equals(value)) {
+					Object index = getPositionOrIndex(entry.getKey(), position, this.position);
+
+					if (type == IndexType.FIRST) {
+						if (this.position)
+							return new Integer[]{(Integer) index};
+						return new String[]{(String) index};
+					}
+
+					indices.add(index);
+				}
+				position++;
+			}
+		} else {
+			for (Object object : objects.getArray(event)) {
+				if (object.equals(value)) {
+					if (type == IndexType.FIRST)
+						return new Integer[]{position};
+					indices.add(position);
+				}
+				position++;
+			}
+		}
+
+		if (indices.isEmpty()) {
+			if (this.position)
+				return new Integer[0];
+			return new String[0];
+		}
+
+		if (type == IndexType.LAST) {
+			if (this.position)
+				return new Integer[]{(Integer) indices.get(indices.size() - 1)};
+			return new String[]{(String) indices.get(indices.size() - 1)};
+		}
+
+		if (this.position)
+			return indices.toArray(new Integer[0]);
+		return indices.toArray(new String[0]);
+	}
+
+	private Object getPositionOrIndex(String key, int position, boolean count) {
+		if (count)
+			return position;
+		return key;
+	}
```

Comment:
this doesn't need to exist, it's only ever used once and doesn't simplify anything, really.